keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1435 @@
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimitPerIndex: 1
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+In this case, we run 10 indexes representing the test suites. We allow for one
+failure per index.
+
+#### Story 2
+
+As a CI/CD platform administrator from the [Story 1](#story-1) I want to be able
+to control the failures with the pod failure policy. In particular, I want
+to be able to use pod failure policy to avoid restarts of some indexes, based
+on exit codes.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimitPerIndex: 1
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+  podFailurePolicy:
+    rules:
+    - action: FailIndex
+      onExitCodes:
+        operator: In
+        values: [42]
+```
+
+#### Story 3
+
+As a CI/CD platform administrator from the [Story 1](#story-1) I want to be able
+to fail the entire Job if the number of failed indexes exceeds 50%. I want to
+do this in order to cut down costs of running the tests in case of compilation
+issues that would result in all tests failing.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimitPerIndex: 1
+  maxFailedIndexes: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### The Job object too big
+
+With the new field `.status.failedIndexes` the Job object can be significantly
+larger as every failed index is recorded in the field.
+
+Note that, the similar risk is also present for Indexed Jobs, regarding the
+already existing `.status.completedIndexes` field (see
+[Indexed Jobs can break with high number of parallelism or completions](https://github.com/kubernetes/kubernetes/issues/118085)).
+
+In order to mitigate this risk we first constrain the `.spec.maxFailedIndexes`
+to `10^5`, which is the same limit as for `.spec.parallelism` currently.
+
+Second, we validate if the fields are inside of the scalability limits:
+1. `.spec.completions<=10^5`, `.spec.parallelism<=10^5`, `spec.maxFailedIndexes<=10^5`
+2. `spec.completions` unlimited (<= max int32 ~2*10^9), `.spec.parallelism<=10^4`, `spec.maxFailedIndexes<=10^4`
```

Comment:
Will those limits be exercised only if backoff-per-index is set?
If so, it sounds fine, if not, that's effectively not-backward compatible, right?

Reply:
Yes, the limits are only enforced with error when `backoffLimitPerIndex` is enabled. 

When `backoffLimitPerIndex` is not set we want to have just a warning above the limits, so that we don't break existing installations.